File: TicTacToe/client-apps/Tic-Tac-Toe-Bevy/src/shared/game_info_receiver.rs
//  Tic-Tac-Toe Bevy Client App
//
//  © 2024 Rust Made Easy. All rights reserved.
//  @author JoelDavisEngineering@Gmail.com

use rumqttc::MqttOptions;
use std::time::Duration;
use tic_tac_toe_rust_client_sdk::models::{EventPlaneConfig, EventPlaneTopicNames};

/// Defines handler for game info events.
pub(crate) trait GameInfoReceiverDelegate {
    fn on_game_ended_in_stalemate();
    fn on_game_ended_in_win();
    fn on_player_added();
    fn on_turn_taken();
}

/// Listens for game state changes published by our Tic Tac Toe service.
pub(crate) struct GameInfoReceiver<T: GameInfoReceiverDelegate> {
    //

    client: rumqttc::Client,
    connection: rumqttc::Connection,
    delegate: T,

    topic_prefix: String,

    topic_game_ended_in_stalemate: String,
    topic_game_ended_in_win: String,
    topic_player_added: String,
    topic_turn_taken: String,
}

impl<T: GameInfoReceiverDelegate> GameInfoReceiver<T> {
    //

    pub(crate) fn new(event_plane_config: EventPlaneConfig, delegate: T) -> Self {
        //

        let (client, connection) = Self::setup_mqtt_client(&event_plane_config);

        let mut instance = Self {
            client,
            connection,
            delegate,
            topic_prefix: event_plane_config.topic_prefix,
            topic_game_ended_in_stalemate: "".to_string(),
            topic_game_ended_in_win: "".to_string(),
            topic_player_added: "".to_string(),
            topic_turn_taken: "".to_string(),
        };

        instance.prebuild_topics();

        instance
    }

    /// Builds a full topic string based on the specified topic name.
    fn build_topic(topic_prefix: &str, topic: EventPlaneTopicNames) -> String {
        format!("{}{}", topic_prefix, topic)
    }

    /// Pre-builds the topics so that we are not parsing each time a message is received.
    fn prebuild_topics(&mut self) {
        self.topic_game_ended_in_stalemate = Self::build_topic(&self.topic_prefix, EventPlaneTopicNames::GameEndedInStalemate);
        self.topic_game_ended_in_win = Self::build_topic(&self.topic_prefix, EventPlaneTopicNames::GameEndedInWin);
        self.topic_player_added = Self::build_topic(&self.topic_prefix, EventPlaneTopicNames::PlayerAdded);
        self.topic_turn_taken = Self::build_topic(&self.topic_prefix, EventPlaneTopicNames::TurnTaken);
    }

    fn setup_mqtt_client(event_plane_config: &EventPlaneConfig) -> (rumqttc::Client, rumqttc::Connection) {
        let mut mqttoptions = MqttOptions::new("", event_plane_config.broker_address.clone(), event_plane_config.broker_port as u16);
        mqttoptions.set_keep_alive(Duration::from_secs(5));
        rumqttc::Client::new(mqttoptions, 10)
    }
}

File: TicTacToe/client-apps/Tic-Tac-Toe-Swift/Tic-Tac-Toe/GameInfoReceiver.swift
//
//  GameInfoReceiver.swift
//
// © 2024 Rust Made Easy. All rights reserved.
// @author JoelDavisEngineering@Gmail.com
//

import Foundation
import MQTTNIO
import OpenAPIClient

/// Defines handler for game info events.
protocol GameInfoReceiverDelegate {    
    func onGameDeleted()
    func onGameEndedInStalemate()
    func onGameEndedInWin()
    func onGameStarted()
    func onPlayerAddedToSession()
    func onSessionDeleted()
    func onTurnTaken()
}

/// Listens for game state changes published by our Tic Tac Toe service.
class GameInfoReceiver {
    
    private var client: MQTTClient!
    private var delegate: GameInfoReceiverDelegate!
    private var eventPlaneConfig: EventPlaneConfig
    
    private var topicGameDeleted: String!
    private var topicGameEndedInStalemate: String!
    private var topicGameEndedInWin: String!
    private var topicGameStarted: String!
    private var topicPlayerAddedToSession: String!
    private var topicSessionDeleted: String!
    private var topicTurnTaken: String!
    
    init(eventPlaneConfig: EventPlaneConfig, delegate: GameInfoReceiverDelegate) {
        self.delegate = delegate
        self.eventPlaneConfig = eventPlaneConfig
        prebuildTopics()
        self.client = setupMqttClient(eventPlaneConfig: eventPlaneConfig)
    }
    
}

extension GameInfoReceiver {
    
    /// Builds a full topic string based on the specified topic name.
    private func buildTopic(topic: EventPlaneTopicNames) -> String {
        String(format: "%@/%@", self.eventPlaneConfig.topicPrefix, topic.rawValue)
    }
    
    /// Pre-builds the topics so that we are not parsing each time a message is received.
    private func prebuildTopics() {
        self.topicGameDeleted = buildTopic(topic: EventPlaneTopicNames.gameDeleted)
        self.topicGameEndedInStalemate = buildTopic(topic: EventPlaneTopicNames.gameEndedInStalemate)
        self.topicGameEndedInWin = buildTopic(topic: EventPlaneTopicNames.gameEndedInWin)
        self.topicGameStarted = buildTopic(topic: EventPlaneTopicNames.gameStarted)
        self.topicPlayerAddedToSession = buildTopic(topic: EventPlaneTopicNames.playerAddedToSession)
        self.topicSessionDeleted = buildTopic(topic: EventPlaneTopicNames.sessionDeleted)
        self.topicTurnTaken = buildTopic(topic: EventPlaneTopicNames.turnTaken)
    }

    /// Initializes the MQTT client and sets up the callbacks used to inform the delegate.
    private func setupMqttClient(eventPlaneConfig: EventPlaneConfig) -> MQTTClient {
                
        let client = MQTTClient(
            configuration: .init(
                target: .host(eventPlaneConfig.brokerAddress, port: eventPlaneConfig.brokerPort)
            ),
            eventLoopGroupProvider: .createNew
        )
        
        client.connect()
        
        let allMessagesForSession = String(format: "%@/#", eventPlaneConfig.topicPrefix)
        client.subscribe(to: allMessagesForSession)

        client.whenConnected { response in
            print("MQTT connected. Is session present: \(response.isSessionPresent)")
        }

        client.whenDisconnected { reason in
            print("MQTT disconnected: \(reason)")
        }

        client.whenMessage { message in
            // Inform the delegate of the event
            switch message.topic {
            case self.topicGameDeleted:
                self.delegate.onGameDeleted()
            case self.topicGameEndedInStalemate:
                self.delegate.onGameEndedInStalemate()
            case self.topicGameEndedInWin:
                self.delegate.onGameEndedInWin()
            case self.topicGameStarted:
                self.delegate.onGameStarted()
            case self.topicPlayerAddedToSession:
                self.delegate.onPlayerAddedToSession()
            case self.topicSessionDeleted:
                self.delegate.onSessionDeleted()
            case self.topicTurnTaken:
                self.delegate.onTurnTaken()
            default:
                print("Warning: GameInfoReceiver - Received unsupported message")
            }
        }
        
        return client
    }
}
